metric/gauge.go

```diff
@@ -0,0 +1,208 @@
+		return entry.(gaugeEntry)
+	} else {
+		entry, _ := g.vals.LoadOrStore(mapKey, g.newEntry())
+		return entry.(gaugeEntry)
+	}
+}
+
+func (g *Gauge) mapKey(labelVals []metricdata.LabelValue) string {
+	vb := &tagencoding.Values{}
+	for _, v := range labelVals {
+		b := make([]byte, 1, len(v.Value)+1)
+		if v.Present {
+			b[0] = 1
+			b = append(b, []byte(v.Value)...)
+		}
+		vb.WriteValue(b)
+	}
+	return string(vb.Bytes())
+}
+
+func (g *Gauge) labelValues(s string) []metricdata.LabelValue {
+	vals := make([]metricdata.LabelValue, 0, len(g.keys))
+	vb := &tagencoding.Values{Buffer: []byte(s)}
+	for range g.keys {
+		v := vb.ReadValue()
+		if v[0] == 0 {
+			vals = append(vals, metricdata.LabelValue{})
+		} else {
+			vals = append(vals, metricdata.NewLabelValue(string(v[1:])))
+		}
+	}
+	return vals
+}
+
+// Float64Gauge represents a float64 value that can go up and down.
+//
+// Float64Gauge maintains a float64 value for each combination of of label values
+// passed to the Set or Add methods.
+type Float64Gauge struct {
+	Gauge
+}
+
+type float64Entry struct {
+	val uint64
+}
+
+func (e *float64Entry) read(t time.Time) metricdata.Point {
+	return metricdata.NewFloat64Point(t, math.Float64frombits(atomic.LoadUint64(&e.val)))
+}
+
+// NewFloat64WithRegistry creates a new gauge with a float64 value.
+func NewFloat64Gauge(name, description string, unit metricdata.Unit, keys ...string) *Float64Gauge {
+	g := &Float64Gauge{
+		Gauge{
+			newEntry: func() gaugeEntry {
+				return new(float64Entry)
+			},
+			keys:  keys,
+			start: time.Now(),
+			desc: metricdata.Descriptor{
+				Name:        name,
+				Description: description,
+				Unit:        unit,
+				LabelKeys:   keys,
+			},
+		},
+	}
+	return g
+}
+
+// Set sets the current gauge value.
+func (g *Float64Gauge) Set(val float64, labelVals ...metricdata.LabelValue) {
+	ge := g.getEntry(labelVals).(*float64Entry)
+	atomic.StoreUint64(&ge.val, math.Float64bits(val))
+}
+
+// Add increments the current gauge value by val.
+func (g *Float64Gauge) Add(val float64, labelVals ...metricdata.LabelValue) {
+	ge := g.getEntry(labelVals).(*float64Entry)
+	var swapped bool
+	for !swapped {
+		oldVal := atomic.LoadUint64(&ge.val)
+		newVal := math.Float64bits(math.Float64frombits(oldVal) + val)
+		swapped = atomic.CompareAndSwapUint64(&ge.val, oldVal, newVal)
+	}
+}
+
+// Int64Gauge represents a int64 gauge value that can go up and down.
+//
+// Int64Gauge maintains an int64 value for each combination of label values passed to the
+// Set or Add methods.
+type Int64Gauge struct {
+	Gauge
+}
+
+type int64GaugeValue struct {
+	val int64
+}
+
+func (v *int64GaugeValue) read(t time.Time) metricdata.Point {
+	return metricdata.NewInt64Point(t, atomic.LoadInt64(&v.val))
+}
+
+// NewInt64Gauge creates a new int64-valued gauge and adds it to the
+// given registry.
+func NewInt64Gauge(name, description string, unit metricdata.Unit, keys ...string) *Int64Gauge {
+	g := &Int64Gauge{
```

Comment:
Shouldn't we keep a reference to the newly created Gauge in order to be able to read/export the value for this Gauge?

Reply:
I envisioned that the Gauge would be added to a metricexport.Registry object as in the example. But I will make it work like Java and probably have a MetricManager that can make Gauges instead of having them be independent objects like this.

Reply:
Done